Rework Axiom Cloudwatch Forwarder (#69)
axiomhq/axiom-cloudwatch-lambda: README.md

# Axiom CloudWatch Forwarder [![CI](https://github.com/axiomhq/axiom-cloudwatch-forwarder/actions/workflows/ci.yaml/badge.svg)](https://github.com/axiomhq/axiom-cloudwatch-forwarder/actions/workflows/ci.yaml)

> Axiom CloudWatch Lambda is an easy-to-use AWS CloudFormation template to send logs from CloudWatch to [Axiom](https://axiom.co). It deploys a Lambda and a subscriber to create the needed CloudWatch log group subscription filters.

Reviewer: ```suggestion
Axiom CloudWatch Forwarder is a set of easy-to-use AWS CloudFormation stacks designed to forward logs from Amazon CloudWatch to [Axiom](https://axiom.co). It includes a Lambda function to handle the forwarding, as well as subscriber and listener stacks to create CloudWatch log group subscription filters for both existing and future log groups.
```

> Axiom CloudWatch Lambda uses the following CloudFormation stacks:

Reviewer: ```suggestion
Axiom CloudWatch Forwarder includes templates for the following CloudFormation stacks:
```

> - Axiom Forwarder creates a Lambda function that ingests logs from CloudWatch and sends them to Axiom.

Reviewer: ```suggestion
- Forwarder creates a Lambda function that forwards logs from CloudWatch to Axiom.
```

> - Subscriber runs once to create subscription filters on the ingest Lambda for all existing CloudWatch log groups.

Reviewer: ```suggestion
- _Subscriber_ runs once to create subscription filters on the _Forwarder_ Lambda for CloudWatch log groups specified by a combination of names, prefix and regular expression.
```

> - LogGroups Listener creates a Lambda function that listens for new log groups and creates subscription filters for them. This way, you don't have to create subscription filters manually for new log groups.

Reviewer: ```suggestion
- _Listener_ creates a Lambda function that listens for new CloudWatch log groups specified by name, pattern, and/or regular expression. It will create subscription filters for them on the _Forwarder_ Lambda. This way, you don't have to create subscription filters manually for new log groups.
```

## Guide

> 1. [Create an Axiom account](https://app.axiom.co).

Reviewer: ```suggestion
1. [Create an Axiom organization](https://app.axiom.co).
```

2. Create a dataset in Axiom.
3. Create an API token in Axiom with permissions to ingest data to the dataset you created.

> 4. [Click this link to launch the Forwarder Stack](https://console.aws.amazon.com/cloudformation/home?#/stacks/new?stackName=axiom-cloudwatch-forwarder&templateURL=https://axiom-cloudformation.s3.amazonaws.com/stacks/axiom-cloudwatch-forwarder-v1.0.0-cloudformation-stack.yaml).

Reviewer: ```suggestion
4. [Click this link to open the Forwarder stack template](https://console.aws.amazon.com/cloudformation/home?#/stacks/new?stackName=axiom-cloudwatch-forwarder&templateURL=https://axiom-cloudformation.s3.amazonaws.com/stacks/axiom-cloudwatch-forwarder-v1.0.0-cloudformation-stack.yaml).
```

Reviewer: Outdated

> 5. Get the created Forwarder lambda ARN from the previous step, and use it to install the Subscriber stack in the next step.

Reviewer: ```suggestion
5. Copy the _Forwarder_ Lambda ARN from the previous step, as it will be referenced in the _Subscriber_ stack.
```

Reviewer: Outdated

> 6. [Click this link to launch the Subscriber stack](https://console.aws.amazon.com/cloudformation/home?#/stacks/new?stackName=axiom-cloudwatch-subscriber&templateURL=https://axiom-cloudformation.s3.amazonaws.com/stacks/axiom-cloudwatch-subscriber-v1.0.0-cloudformation-stack.yaml) and automatically subscribe to all existing log groups. You can filter the log groups by one or a combination of names, regex pattern and a prefix.

Reviewer: ```suggestion
6. [Click this link to open the Subscriber stack template](https://console.aws.amazon.com/cloudformation/home?#/stacks/new?stackName=axiom-cloudwatch-subscriber&templateURL=https://axiom-cloudformation.s3.amazonaws.com/stacks/axiom-cloudwatch-subscriber-v1.0.0-cloudformation-stack.yaml).
```

How this works now handled in description above.

> 7. [Click this link to automatically subscribe to new log groups](https://console.aws.amazon.com/cloudformation/home?#/stacks/new?stackName=axiom-cloudwatch-log-groups-listener&templateURL=https://axiom-cloudformation.s3.amazonaws.com/stacks/axiom-cloudwatch-log-groups-listener-v1.0.0-cloudformation-stack.yaml).

Reviewer: ```suggestion
7. [Click this link to open the Listener stack template](https://console.aws.amazon.com/cloudformation/home?#/stacks/new?stackName=axiom-cloudwatch-log-groups-listener&templateURL=https://axiom-cloudformation.s3.amazonaws.com/stacks/axiom-cloudwatch-log-groups-listener-v1.0.0-cloudformation-stack.yaml).
```

Reviewer: Outdated

## Removing subscription filters

> If later on, you wanted to remove subscription filters for one or multiple log groups, you can [launch the Unsubscriber stack](https://console.aws.amazon.com/cloudformation/home?#/stacks/new?stackName=axiom-cloudwatch-subscriber&templateURL=https://axiom-cloudformation.s3.amazonaws.com/stacks/axiom-cloudwatch-unsubscriber-v1.0.0-cloudformation-stack.yaml)

Reviewer: ```suggestion
If later on you want to remove subscription filters for one or more log groups, you can [click this link to open the Unsubscriber stack template](https://console.aws.amazon.com/cloudformation/home?#/stacks/new?stackName=axiom-cloudwatch-subscriber&templateURL=https://axiom-cloudformation.s3.amazonaws.com/stacks/axiom-cloudwatch-unsubscriber-v1.0.0-cloudformation-stack.yaml)
```

Reviewer: Covered in incoming commit.

> The log group filtering works the same way as the subscriber stack. You can filter the log groups by one or a combination of names, regex pattern and a prefix.

Reviewer: ```suggestion
The log group filtering works the same way as the _Subscriber_ stack. You can filter the log groups by a combination of names, prefix and regular expression.
```

Reviewer: Covered in incomig commit

## Filtering CloudWatch log groups

> The Subscriber and Unsubscriber stacks allow you to filter the log groups by one or a combination of names, regex pattern and a prefix. Basically,

Reviewer: ```suggestion
The _Subscriber_ and _Unsubscriber_ stacks allow you to filter the log groups by one or a combination of names, prefix and regular expression.
```

> You can whitelist a number of log groups. If you don't specify any filter, the stack will subscribe to all log groups. If a log group matches one of the filters, it will be subscribed to.

Reviewer: ```suggestion
You can allowlist specific log groups. If no filters are specified, the stacks will subscribe to or unsubscribe from all log groups.
```

> **Example**

Reviewer: For completeness, would be nice to include an example of regular expression.

Author: will do.

Author: done

Let's assume we have this list of log groups:

```

> /aws/lambda/functionFoo
> /aws/lambda/functionBar
> /aws/eks/cluster/cluster1
> /aws/rds/instanceFoo

Reviewer: ```suggestion
/aws/lambda/function-foo
/aws/lambda/function-bar
/aws/eks/cluster/cluster-1
/aws/rds/instance-baz
```

```

> - To subscribe to the lambda log groups only, a prefix filter with the value of `/aws/lambda` would do the job.

Reviewer: ```suggestion
- To subscribe to the Lambda log groups exclusively, a prefix filter with the value of `/aws/lambda` would do the job.
```

> - To subscribe to eks and rds log groups, a list of names with the value of `/aws/eks/cluster/cluster1,/aws/rds/instanceFoo` would do the job.

Reviewer: ```suggestion
- To subscribe to EKS and RDS log groups, a list of names with the value of `/aws/eks/cluster/cluster-1,/aws/rds/instance-foo` would work well.
```

> - To subscribe to the eks log group and all lambda log groups, a combination of prefix and names list would select them.

Reviewer: ```suggestion
- To subscribe to the EKS log group and all Lambda log groups, a combination of prefix and names list would select them. Log group names, prefixes, and regular expressions are additive, meaning the union of all provided inputs will be matched.
```

## Log Groups Listener architecture

The Log Groups Listener stack does the following:

> - It creates an S3 bucket for Cloudtrail.

Reviewer: ```suggestion
- It creates an Amazon S3 bucket for AWS CloudTrail.
```
